homeassistant/components/fan/knx.py

```diff
@@ -0,0 +1,142 @@
+CONF_SPEED_STATE_ADDRESS = 'speed_state_address'
+
+DEFAULT_NAME = 'KNX Fan'
+DEPENDENCIES = ['knx']
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_SPEED_ADDRESS): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_SPEED_STATE_ADDRESS): cv.string,
+})
+
+
+async def async_setup_platform(hass,
+                               config,
+                               async_add_devices,
+                               discovery_info=None):
+    """Set up fan(s) for KNX platform."""
+    if discovery_info is not None:
+        async_add_devices_discovery(hass, discovery_info, async_add_devices)
+    else:
+        async_add_devices_config(hass, config, async_add_devices)
+
+
+@callback
+def async_add_devices_discovery(hass, discovery_info, async_add_devices):
+    """Set up fans for KNX platform configured via xknx.yaml."""
+    entities = []
+    for device_name in discovery_info[ATTR_DISCOVER_DEVICES]:
+        device = hass.data[DATA_KNX].xknx.devices[device_name]
+        entities.append(KNXFan(hass, device))
+    async_add_devices(entities)
+
+
+@callback
+def async_add_devices_config(hass, config, async_add_devices):
+    """Set up fan for KNX platform configured within plattform."""
+    import xknx
+    fan = xknx.devices.Fan(
+        hass.data[DATA_KNX].xknx,
+        name=config.get(CONF_NAME),
+        group_address_speed=config.get(CONF_SPEED_ADDRESS),
+        group_address_speed_state=config.get(CONF_SPEED_STATE_ADDRESS))
+    hass.data[DATA_KNX].xknx.devices.add(fan)
+    async_add_devices([KNXFan(hass, fan)])
+
+
+class KNXFan(FanEntity):
+    """Representation of a KNX fan."""
+
+    def __init__(self, hass, device):
+        """Initialize the entity."""
+        self.device = device
+        self.hass = hass
+        self.async_register_callbacks()
+
+    @callback
+    def async_register_callbacks(self):
+        """Register callbacks to update hass after device was changed."""
+        async def after_update_callback(device):
+            """Call after device was updated."""
+            # pylint: disable=unused-argument
+            await self.async_update_ha_state()
+        self.device.register_device_updated_cb(after_update_callback)
+
+    @property
+    def name(self):
+        """Return the name of the KNX device."""
+        return self.device.name
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self.device.current_speed() != 0
+
+    @property
+    def should_poll(self):
+        """No polling needed within KNX."""
+        return False
+
+    @property
+    def speed(self) -> str:
+        """Return the current speed."""
+        if self.device.current_speed() is not None:
+            return self.device.current_speed()
+
+    @property
+    def speed_list(self) -> list:
+        """Get the list of available speeds."""
+        return [0, 25, 50, 75, 100]
```

Comment:
Please extract this into a constant and return that here. At the moment we only allow the speed constants that are defined in the base fan component:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/fan/__init__.py#L43-L46

We can use two dictionaries to convert between home assistant speed and knx speed.

This might change in the future though:
https://github.com/home-assistant/architecture/issues/27

Reply:
I tried first to do some mapping but the KNX fan has 5 pre-set speeds (if including 0 -> OFF) and there are only 4 constants for the base fan.
Thats even more problematic if you use a really advanded KNX fan since the value is a percentage (0 -> 100%).

Just had a look at the architecture issue you pointed out... I think im stuck until its "fixed". Some trick for me ?

Reply:
You could skip one of the knx speeds going from home assistant to knx, and map two speeds (including the skipped one) when going from knx to home assistant. Eg:

```py
HA_STATE_TO_KNX = {
    SPEED_OFF: 0,
    SPEED_LOW: 25,
    SPEED_MEDIUM: 50,
    SPEED_HIGH: 100,
}

KNX_STATE_TO_HA = {
    0: SPEED_OFF,
    25: SPEED_LOW,
    50: SPEED_MEDIUM,
    75: SPEED_MEDIUM,
    100: SPEED_HIGH,
}
```

Reply:
But then my fan (and others) will not work satisfyingly... Speed nr 3 (75%) is missing in your example...
As KNX devices have their own internal logic and many of them are automated (semi-automated), if the system chooses speed 75%, then HASS will display 100%

Reply:
If we need to translate the full percent range we could do something like this:
```py
DISCRETE_TO_HA_STATE = {
    0: SPEED_OFF,
    1: SPEED_LOW,
    2: SPEED_MEDIUM,
    3: SPEED_HIGH,
}

def translate_prc(prc):
    prc_span = 100
    discrete_span = 3
    prc_scaled = prc / prc_span
    discrete_val = math.ceil(prc_scaled * discrete_span)
    return DISCRETE_TO_HA_STATE[discrete_val]
```

Reply:
Would it be possible to get the speed list from the configuration ?
Default: we use what you suggest
Optional: speed are provided using the configuration of the component ?

Reply:
No the speed list should be fixed.

Reply:
MQTT is an exception and the wink implementation is not correct.